driver/src/main/com/mongodb/client/gridfs/GridFSBucket.java

```diff
@@ -0,0 +1,253 @@
+/*
+ * Copyright 2015 MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.mongodb.client.gridfs;
+
+import com.mongodb.ReadPreference;
+import com.mongodb.WriteConcern;
+import com.mongodb.annotations.ThreadSafe;
+import com.mongodb.client.gridfs.model.GridFSDownloadByNameOptions;
+import com.mongodb.client.gridfs.model.GridFSUploadOptions;
+import org.bson.BsonValue;
+import org.bson.conversions.Bson;
+import org.bson.types.ObjectId;
+
+import java.io.InputStream;
+import java.io.OutputStream;
+
+/**
+ * Represents a GridFS Bucket
+ *
+ * @since 3.1
+ */
+@ThreadSafe
+public interface GridFSBucket {
+
+    /**
+     * Sets the chunk size in bytes. Defaults to 255.
+     *
+     * @return the chunk size in bytes.
+     */
+    int getChunkSizeBytes();
+
+    /**
+     * Get the write concern for the GridFSBucket.
+     *
+     * @return the {@link com.mongodb.WriteConcern}
+     */
+    WriteConcern getWriteConcern();
+
+    /**
+     * Get the read preference for the GridFSBucket.
+     *
+     * @return the {@link com.mongodb.ReadPreference}
+     */
+    ReadPreference getReadPreference();
+
+    /**
+     *  Create a new GridFSBucket instance with a new chunk size in bytes.
+     *
+     * @param chunkSizeBytes the new chunk size in bytes.
+     * @return a new GridFSBucket instance with the different chunk size in bytes
+     */
+    GridFSBucket withChunkSizeBytes(int chunkSizeBytes);
+
+    /**
+     * Create a new GridFSBucket instance with a different read preference.
+     *
+     * @param readPreference the new {@link ReadPreference} for the database
+     * @return a new GridFSBucket instance with the different readPreference
+     */
+    GridFSBucket withReadPreference(ReadPreference readPreference);
+
+    /**
+     * Create a new GridFSBucket instance with a different write concern.
+     *
+     * @param writeConcern the new {@link WriteConcern} for the database
+     * @return a new GridFSBucket instance with the different writeConcern
+     */
+    GridFSBucket withWriteConcern(WriteConcern writeConcern);
+
```

Comment:
This was removed in a previous iteration - as we don't have a withDatabaseName in MongoDatabase or similar in MongoCollection.

Done.